gc.rb

```diff
@@ -130,33 +130,59 @@ def self.count
   #
   #  The hash includes information about internal statistics about GC such as:
   #
-  #      {
-  #          :count=>0,
-  #          :heap_allocated_pages=>24,
-  #          :heap_sorted_length=>24,
-  #          :heap_allocatable_pages=>0,
-  #          :heap_available_slots=>9783,
-  #          :heap_live_slots=>7713,
-  #          :heap_free_slots=>2070,
-  #          :heap_final_slots=>0,
-  #          :heap_marked_slots=>0,
-  #          :heap_eden_pages=>24,
-  #          :heap_tomb_pages=>0,
-  #          :total_allocated_pages=>24,
-  #          :total_freed_pages=>0,
-  #          :total_allocated_objects=>7796,
-  #          :total_freed_objects=>83,
-  #          :malloc_increase_bytes=>2389312,
-  #          :malloc_increase_bytes_limit=>16777216,
-  #          :minor_gc_count=>0,
-  #          :major_gc_count=>0,
-  #          :remembered_wb_unprotected_objects=>0,
-  #          :remembered_wb_unprotected_objects_limit=>0,
-  #          :old_objects=>0,
-  #          :old_objects_limit=>0,
-  #          :oldmalloc_increase_bytes=>2389760,
-  #          :oldmalloc_increase_bytes_limit=>16777216
-  #      }
+  #  [count]
+  #    The total number of garbage collections run since process start,
+  #    including minor and major garbage collections
+  #  [heap_allocated_pages]
+  #    The total number of `:heap_eden_pages` + `:heap_tomb_pages`
+  #  [heap_sorted_length]
+  #    The number of pages that can fit into memory between the first and last
+  #    page
+  #  [heap_allocatable_pages]
+  #    The total number of pages the process could allocate without additional GC
+  #  [heap_available_slots]
+  #    The total number of slots in all `:heap_allocated_pages`
+  #  [heap_live_slots]
+  #    The total number of slots which contain live objects
+  #  [heap_free_slots]
+  #    The total number of slots which do not contain live objects
+  #  [heap_final_slots]
+  #    The total number of slots with finalizers
+  #  [heap_marked_slots]
+  #    The total number of old objects and write barrier unprotected objects
+  #  [heap_eden_pages]
+  #    The total number of pages which contain at least one live slot
+  #  [heap_tomb_pages]
+  #    The total number of pages which do not contain any live slots
+  #  [total_allocated_pages]
+  #    The cumulative number of pages allocated since process start
+  #  [total_freed_pages]
+  #    The cumulative number of pages freed since process start
+  #  [total_allocated_objects]
+  #    The cumulative number of objects allocated since process start
+  #  [total_freed_objects]
+  #    The cumulative number of objects freed since process start
+  #  [malloc_increase_bytes]
+  #    Amount of memory allocated on heap for live, young objects
+  #  [malloc_increase_bytes_limit]
+  #    When `:malloc_increase_bytes` crosses this limit, minor GC is triggered
```

Comment:
I'm not too familiar with this part, but looking at the code, it seems like `malloc_increase_bytes` is incremented with the allocation size in [Ruby's wrapper for `malloc`](https://github.com/ruby/ruby/blob/745287d43a8fb63c84be986b23319d40e6affe2f/gc.c#L11424) and decremented in [Ruby's wrapper for `free`](https://github.com/ruby/ruby/blob/745287d43a8fb63c84be986b23319d40e6affe2f/gc.c#L11708). It looks like GC is triggered when `malloc_increase_bytes_limit` is hit. GC will cause `malloc_increase_bytes_limit` to grow accordingly.

Reply:
Do you think that "allocated on the heap for live, young objects" captures that it's incremented on the malloc wrapper and decremented on the `free` wrapper?

When we read the code and looked at results of `GC.stat`, we were pretty sure that this number does not include old objects.

Reply:
They [both seem to be incremented and decremented together](https://github.com/ruby/ruby/blob/745287d43a8fb63c84be986b23319d40e6affe2f/gc.c#L11326).

Additionally, we can do the following experiment:

```ruby
obj = Object.new
arr = []

# Ensure that all objects are promoted to old by triggering GC
10.times do
  GC.start
end

before = GC.stat(:malloc_increase_bytes)

100.times do
  arr << obj
end

after = GC.stat(:malloc_increase_bytes)

puts before, after
```

I get output that looks like this:

```
1104
2136
```

Which grows by about 1000 bytes by the allocation for the array resize.

Reply:
Hi @peterzhu2118! I'm not sure I follow the change you're suggesting here (or below with `old_malloc_increase_bytes`. Is it to indicate that this limit will change with GC?

Reply:
Ruby's GC wraps the C library functions `malloc`, `realloc`, `free` with its own wrappers that will try to GC to free memory when `malloc` or `realloc` fails (i.e. out of memory). It looks like one of the metrics for determining when to GC is the `malloc_increase_bytes` and `old_malloc_increase_bytes` values used to keep track of the number of bytes allocated through system (`malloc`).

Reply:
Hmmm... I see that they are incremented in the same place, but then how do the end up as different values? Modifying your script slightly to show oldmalloc, we see that the values are not the same:

```ruby
obj = Object.new
arr = []

# Ensure that all objects are promoted to old by triggering GC
10.times do
  GC.start
end

before = GC.stat(:malloc_increase_bytes)
before_old = GC.stat(:oldmalloc_increase_bytes)

100.times do
  arr << obj
end

after = GC.stat(:malloc_increase_bytes)
after_old = GC.stat(:oldmalloc_increase_bytes)

puts before, after
puts before_old, after_old
```

Reply:
I'm pretty sure there's a bug in Ruby 😄. I opened #4860.

The values for [`malloc_increase_bytes`](https://github.com/ruby/ruby/blob/b45fe48fbb20e334ef6dd1a85d5b7a37e6701d40/gc.c#L8888) and [`oldmalloc_increase_bytes`](https://github.com/ruby/ruby/blob/b45fe48fbb20e334ef6dd1a85d5b7a37e6701d40/gc.c#L8938) are reset in `gc_reset_malloc_info`, which [is ran when GC is triggered](https://github.com/ruby/ruby/blob/b45fe48fbb20e334ef6dd1a85d5b7a37e6701d40/gc.c#L9074). Note that `oldmalloc_increase_bytes` SHOULD only be reset on major GC, but because of the bug it's incorrect since Ruby 2.2.

Reply:
Ohhhhh! Nice catch! So we should describe `malloc_increase_bytes` as something like "Amount of memory allocated on the heap for objects. Decreased by any GC" and "Decreased by major GC" for oldmalloc_increase_bytes?

Reply:
I updated the comments to this - let me know how it looks!